src/modalities/models/coca/multi_modal_decoder.py

```diff
@@ -0,0 +1,123 @@
+
+        if activation == ActivationType.GELU:
+            mlp = partial(MLP, in_features=n_embd, hidden_features=ffn_hidden, bias=bias, dropout=dropout)
+        elif activation == ActivationType.FUSED_SWIGLU:
+            mlp = partial(xops.SwiGLU, in_features=n_embd, hidden_features=ffn_hidden, bias=bias)
+        else:
+            raise NotImplementedError(f"activation type {activation} not implemented")
+
+        self.ln_1 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+        self.attn = MultiHeadAttention(
+            n_embd=n_embd, n_head=n_head, bias=bias, attention_config=attention_config, attention_type=attention_type
+        )
+
+        if not self.with_context or self.add_extra_mlp:
+            self.ln_2 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+            self.mlp = mlp()
+
+        if self.with_context:
+            self.ln_3 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+            self.cross_attn = MultiHeadAttention(
+                n_embd=n_embd,
+                n_head=n_head,
+                bias=bias,
+                attention_config=attention_config,
+                attention_type=AttentionType.CROSS_ATTENTION,
+            )
+            self.ln_4 = LayerNorm(ndim=n_embd, bias=bias, epsilon=epsilon)
+            self.mlp_2 = mlp()
+
+    def forward(self, x: torch.Tensor, context: torch.Tensor = None) -> torch.Tensor:
+        x = x + self.attn(self.ln_1(x))
+        if not self.with_context or self.add_extra_mlp:
+            x = x + self.mlp(self.ln_2(x))
+        if self.with_context:
+            x = x + self.cross_attn(self.ln_3(x), context=context)
+            x = x + self.mlp_2(self.ln_4(x))
+        return x
+
+
+class MultiModalTextDecoder(NNModel):
+    def __init__(
+        self,
+        sample_key: str,
+        prediction_key: str,
+        block_size: int,
+        vocab_size: int,
+        n_layer: int,
+        n_head: int,
+        n_embd: int,
+        ffn_hidden: int,
+        dropout: float,
+        bias: bool,
+        activation: ActivationType,
+        epsilon: float,
+        attention_config: AttentionConfig,
+    ):
+        super().__init__()
+        self.sample_key = sample_key
+        self.prediction_key = prediction_key
+        self.block_size = block_size
+
+        self.transformer = nn.ModuleDict(
+            dict(
+                h=nn.ModuleList(
+                    [
+                        TransformerBlock(
+                            n_embd=n_embd,
+                            bias=bias,
+                            epsilon=epsilon,
+                            activation=activation,
+                            n_head=n_head,
+                            dropout=dropout,
+                            ffn_hidden=ffn_hidden,
+                            with_context=True,
+                            attention_type=AttentionType.CAUSAL_SELF_ATTENTION,
+                            attention_config=attention_config,

```

Comment:
```suggestion
                            attention_config=attention_config,
                            add_extra_mlp=False,
```

Reply:
It might be useful to explicitly specify `add_extra_mlp` here (instead of relying on the default value), since `with_context` is `True` (line 105) which makes `add_extra_mlp` relevant.